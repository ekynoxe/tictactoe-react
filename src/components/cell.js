import React from 'react';
import { connect } from 'react-redux';
import actions from '../actions';

class BaseCell extends React.Component {
    select() {
        this.props.dispatch(actions.selectCell(this.props.index));
    }

    render() {
        let mark;

        if (this.props.mark) {
            mark = this.props.mark;

        } else {
            mark = <button onClick={ this.select.bind(this) }>&nbsp;</button>;
        }

        return(<div className='cell'><span className='mark'>{ mark }</span></div>);
    }
}

export const Cell = connect()(BaseCell);
