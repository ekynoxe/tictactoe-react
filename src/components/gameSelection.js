import React from 'react';
import { connect } from 'react-redux';
import gameTypes from '../types';
import actions from '../actions';

export class GameSelection extends React.Component {
    selectGame(type) {
        this.props.dispatch(actions.selectGame(type));
    }

    render() {
        return <div className='buttons'>
            <button key='btn_single_player' className='button' onClick={ this.selectGame.bind(this, gameTypes.singlePlayer) }>Single player</button>
            <button key='btn_two_players_local' className='button' onClick={ this.selectGame.bind(this, gameTypes.twoPlayersLocal) }>Two players in this window</button>
        </div>;
    }
}

export default connect()(GameSelection);
