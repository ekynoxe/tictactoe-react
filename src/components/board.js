import React from 'react';
import { Cell } from './cell';

export class Board extends React.Component {
    render() {
        let rows = [];
        let cells;

        this.props.board.forEach( (m, i) => {
            if (!(i % 3)) {
                cells = [];
            }

            cells.push(<Cell key={ `c_${i}` } mark={ m } index={ i } />);

            if (!((i + 1) % 3)) {
                let row = Math.floor(i / 3);
                rows.push(<div key={ `r_${row}` } className='row'>{ cells }</div>);
            }
        });

        return(<div className='board'>{ rows }</div>);
    }
}
