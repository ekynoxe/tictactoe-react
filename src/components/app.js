import React from 'react';
import { connect } from 'react-redux';
import { Board } from './board';
import actions from '../actions';
import players from '../players';
import states from '../states';

class BaseApp extends React.Component {
    reset() {
        this.props.dispatch(actions.reset());
    }

    setPlayer(player) {
        this.props.dispatch(actions.setPlayer(player));
    }

    componentWillUpdate(nextProps) {
        // On first load or reset, no player will be selected, so we randomly
        //  select a player to start the game here.
        if (!nextProps.currentPlayer) {
            this.setPlayer(Math.round(Math.random()) === 0 ? players.o : players.x);
        }
    }

    render() {
        // Waiting for application to load the local store
        if (!this.props.rehydrated) {
            return <div>Loading...</div>;
        }

        let notice;

        if (states.inplay !== this.props.gameState) {
            notice = 'Game over (declare winner here...)';

        } else if (this.props.currentPlayer) {
            notice = `Current player: ${ this.props.currentPlayer.toUpperCase() }`;
        }

        return(<div>
            <Board board={ this.props.board } />
            <p className='notice'>{ notice }</p>
            <button onClick={ this.reset.bind(this) }>Restart</button>
        </div>);
    }
}

function mapStateToProps(state) {
    return {
        board: state.board,
        currentPlayer: state.currentPlayer,
        gameState: state.gameState
    };
}

export const App = connect(mapStateToProps)(BaseApp);
