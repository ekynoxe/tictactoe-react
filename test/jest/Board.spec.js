/* global jest, test, expect */
import React from 'react';
import renderer from 'react-test-renderer';
import { Board } from '../../src/components/board';
import { inPlayBoard } from '../boards';

jest.mock('../../src/components/cell', () => 'Cell');

test('Board matches snapshot', () => {
    const component = renderer.create(
        <Board board={ inPlayBoard } />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
