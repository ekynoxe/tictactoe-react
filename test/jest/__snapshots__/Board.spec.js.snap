// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board matches snapshot 1`] = `
<div
  className="board"
>
  <div
    className="row"
  >
    <Cell
      index={0}
      mark="x"
    />
    <Cell
      index={1}
      mark="o"
    />
    <Cell
      index={2}
      mark="x"
    />
  </div>
  <div
    className="row"
  >
    <Cell
      index={3}
      mark="o"
    />
    <Cell
      index={4}
      mark="o"
    />
    <Cell
      index={5}
      mark="x"
    />
  </div>
  <div
    className="row"
  >
    <Cell
      index={6}
      mark={null}
    />
    <Cell
      index={7}
      mark={null}
    />
    <Cell
      index={8}
      mark={null}
    />
  </div>
</div>
`;
